\newif\ifmbeformat
\mbeformatfalse

\ifmbeformat
	\documentclass[nogrid]{MBE}%
\else
	\documentclass[twocolumn]{article}
	\usepackage{geometry}
	\newgeometry{
		top=1in,
		bottom=1in,
		outer=1in,
		inner=1in,
	}
	\usepackage{natbib}
	\renewcommand{\baselinestretch}{1.25}		
	\usepackage{authblk}
\fi

\usepackage[normalem]{ulem}
\usepackage{xcolor} 
\newcommand{\revch}[1]{{\color{blue} #1}}
\newcommand{\revcom}[1]{{\color{orange} #1}}
\newcommand{\revrem}[1]{{\color{red} \sout{#1}}}

\usepackage{url}
\usepackage{cuted}
\usepackage{amsmath}
\usepackage{subfig}
%\usepackage{caption}
\usepackage{tabularx}
\usepackage{longtable}
\usepackage{caption}
\usepackage{verbatim}
\usepackage{lipsum}
\usepackage{enumerate}
\usepackage{enumitem}
%load any additional packages
\usepackage{amssymb}
\usepackage{scalerel,stackengine}
\stackMath
\newcommand\reallywidehat[1]{%
	\savestack{\tmpbox}{\stretchto{%
			\scaleto{%
				\scalerel*[\widthof{\ensuremath{#1}}]{\kern-.6pt\bigwedge\kern-.6pt}%
				{\rule[-\textheight/2]{1ex}{\textheight}}%WIDTH-LIMITED BIG WEDGE
			}{\textheight}% 
		}{0.5ex}}%
	\stackon[1pt]{#1}{\tmpbox}%
}
\usepackage{amsfonts}
\usepackage{mathtools}
\usepackage{multirow}
\usepackage{bibentry}
%\usepackage{epstopdf} 
%\usepackage{xr}
%\usepackage[landscape,a4paper]{geometry}
\usepackage{booktabs} 
\usepackage{colortbl} 
%\usepackage{xcolor} 
%\usepackage{xfrac}
%\newcommand{ra}[1]{renewcommand{arraystretch}{#1}}
%\usepackage{pdflscape}
%\usepackage{longtable}
%\usepackage{stackrel}
\usepackage{hyperref}
%\usepackage{graphicx}
%\usepackage{svg}
\usepackage{setspace}
%\usepackage{refcheck}
%\usepackage{listings}
%\usepackage{color} 
\usepackage{array}
\usepackage{xr}
\externaldocument[supplementary-]{supplementary}


\newif\iffigures
\figurestrue

\ifmbeformat
\jshort{mst}

\volname{}

\jvolume{0}

\jvol{}

\jissue{0}

\pubyear{2013}

\mstype{Article}

\artid{012}

\access{Advance Access publication March 3, 2013}
\fi

\newcommand{\abstractext}{We present a probabilistic model of protein evolution that captures several important features of protein sequence and local structure. The key feature being  dependencies between neighbouring amino acid positions that are temporal in nature due to sequence mutations that occur during evolution. The model is trained on a large number of protein alignments and corresponding phylogenetic trees that represent the evolutionary history of the aligned proteins. This yields a model that acts as a rich prior distribution over protein evolution that can be used to perform several important inference tasks. One such task being Bayesian reconstruction of ancestral virus protein sequences, which we demonstrate to have better accuracy than competing methods. The model provides a complete probabilistic description of each protein's backbone structure using an angle and bond length representation. Structure evolution is modelled jointly with sequence, permitting ancestral structures and sequences to be reconstructed in a phylogenetically rigorous manner. Likewise, the model can perform homology modelling to predict the unknown local backbone structure of a known protein sequence using additional information from potentially large numbers of homologous proteins.  The model is highly flexible with respect to input, implying that arbitrary combinations of protein sequences and structures can be used when performing various inference tasks. The current model does not capture global features of protein structure that are necessary for accurate homology modelling or reconstruction of ancestral three-dimensional structures. However, it is ultimately expected to be combined with protein structure prediction models that account for such long-range dependencies, but that do not account for evolutionary information that can substantially enhance predictions of structures.}

\begin{document}

\title{Reconstructing ancestral protein	sequences and structures}

\ifmbeformat
\author[Golden et al.]{Michael \surname{Golden},$^{\ast,1}$, Jotun Hein,$^{2}$ Thomas Hamerlyck,$^{3}$ and Oliver Pybus,$^{1}$}

\address{
$^{1}$Department of Zoology, University of Oxford, UK\\
$^{2}$Department of Statistics, University of Oxford, UK\\
$^{3}$Bioinformatics Centre, Section for Computational and RNA Biology, Department of Biology and Image Section, Department of Computer Science, University of Copenhagen\\
}

\history{Received 13 July 2017; reviews returned 26 November 2017; accepted 30 November 2017}

\coresp{E-mail: golden@phylo.dev}

%\datade{Protein families were obtained from the HOMSTRAD database.}

\editor{Name Surname}

\else
\author[1]{Michael Golden}
\author[2]{Thomas Hamelryck}
\author[1]{Oliver Pybus}
\affil[1]{Department of Zoology, University of Oxford, UK}
%\affil[2]{Department of Statistics, University of Oxford, UK}
\affil[2]{Bioinformatics Centre, Section for Computational and RNA Biology, Department of Biology and Image Section, Department of Computer Science, University of Copenhagen}
\fi


\ifmbeformat
\abstract{\abstractext}
\keyword{Evolution, protein structure, probabilistic model}
\maketitle

\else

\onecolumn
\maketitle
\begin{abstract}
\normalsize
\abstractext
\end{abstract}
\twocolumn
\fi



\section{Introduction}

%%% Methods %%%
\section{Methods}

\subsection{Model}
\begin{figure*}
	\centering
	\includegraphics[width=2.0\columnwidth]{figures/model-structure-final.pdf}
	\caption{Above: a depiction of a protein backbone (three amino acids long) with the $\omega$, $\phi$ and $\psi$ dihedral angles and the three additional bond angles ($\tau^{(1)}_{i}$, $\tau^{(2)}_{i}$, $\tau^{(3)}_{i}$) shown. Bond angles and bond lengths are not to scale. Also shown are $\text{C}_{\alpha}$ atoms which attach to the amino acid side-chains. Each amino acid side-chain determines the characteristic nature of each amino acid. Every amino acid position corresponds to a hidden node in the model below.\newline
	Below: Graphical depiction of the model architecture showing three amino acid positions ($i-1$, $i$, and $i+1$) at two time instants ($t_0$ and $t_1$) along a single branch of a phylogenetic tree.}%
	\label{fig:model-structure}%
\end{figure*}

\subsubsection{Model of a single protein}
A single protein consisting of $n$ amino acids:
\begin{align*}
 P_a  & = (H_a, S_a,X_a)\\
 & =\langle (H^{1}_a, S^{1}_a,X^{1}_a),\ldots,(H^{n}_a, S^{n}_a,X^{n}_a) \rangle
\end{align*}
is a sequence of aligned sites where each site $i$ is associated with a discrete-valued hidden state, $H_a^{i}$ (taking on one of $h$ possible values), a discrete-valued amino acid observation, $S_a^{i}$ (representing one of the twenty possible amino acids), and a corresponding vector of continuous-valued structural observations $X_a^{i}$ representing the backbone structure of the protein.

\paragraph{Structural observations}
The set of structural observations, $X_a^{i}$, at a particular site, $i$, consists of nine continuous-valued variables: three dihedral angles ($\phi_{i},\psi_{i},\omega_{i}$), three additional bond angles ($\tau^{(1)}_{i}=\reallywidehat{C{\alpha}_{i-1},C_{i-1},N_{i}}$, $\tau^{(2)}_{i}=\reallywidehat{C_{i-1},N_{i},C{\alpha}_{i}}$, $\tau^{(3)}_{i}=\reallywidehat{N_{i},C{\alpha}_{i},C_{i}}$), and three bond lengths ($b_{i}^{(1)}=\overrightarrow{C_{i-1},N_{i}}$, $b_{i}^{(2)}=\overrightarrow{N_{i},C{\alpha}_{i}}$, $b_{i}^{(3)}=\overrightarrow{C{\alpha}_{i},C_{i}}$). 

Note that $\phi_{1}$, $\tau^{(1)}_{1}$, and $\tau^{(2)}_{1}$, are undefined at the first position in the peptide backbone of an unaligned protein. Similarly, $\psi_{n}$ and  $\omega_{n}$ are undefined for the last position, $n$, in each unaligned protein.

Given the structural observations, $X_{a}$, it is possible to exactly reconstruct the three-dimensional coordinates of each atom in a protein's backbone \citep{Parsons2005}.

The dihedral angle observations  ($\omega_{i},\phi_{i},\psi_{i}$) at a site, $i$, are each distributed according a univariate von Mises (vM) distribution with mean $\mu$ and concentration parameter $\kappa$ conditional on the hidden state $H_{a}^{i}$ and the amino acid $S_{a}^{i}$:
\begin{equation}
\label{eq:dihedral_density}
\begin{array}{ccc}
\omega_{i} & \sim & \text{vM}\big(\mu_{\omega}(H_{a}^{i},S_{a}^{i}),\,\ensuremath{\kappa_{\omega}(H_{a}^{i},S_{a}^{i})}\big)
\\
\phi_{i} & \sim & \text{vM}\big(\mu_{\phi}(H_{a}^{i},S_{a}^{i}),\,\ensuremath{\kappa_{\phi}(H_{a}^{i},S_{a}^{i})}\big)
\\
\psi_{i} & \sim & \text{vM}\big(\mu_{\psi}(H_{a}^{i},S_{a}^{i}),\,\ensuremath{\kappa_{\psi}(H_{a}^{i},S_{a}^{i})}\big)
\end{array}
\end{equation}

The three additional bond angles ($\tau^{(1)}_{i}$, $\tau^{(2)}_{i}$, $\tau^{(3)}_{i}$) are each distributed according a univariate (vM) distribution conditional on the hidden state $H_{a}^{i}$:
\begin{equation}
\label{eq:angle_density}
\begin{array}{ccc}
\tau^{(1)}_{i} & \sim & \text{vM}\big(\mu_{\tau^{(1)}}(H_{a}^{i}),\,\ensuremath{\kappa_{\tau^{(1)}}(H_{a}^{i})}\big)
\\
\tau^{(2)}_{i} & \sim & \text{vM}\big(\mu_{\tau^{(2)}}(H_{a}^{i}),\,\ensuremath{\kappa_{\tau^{(2)}}(H_{a}^{i})}\big)
\\
\tau^{(3)}_{i} & \sim & \text{vM}\big(\mu_{\tau^{(3)}}(H_{a}^{i}),\,\ensuremath{\kappa_{\tau^{(3)}}(H_{a}^{i})}\big)
\end{array}
\end{equation}

The three bond lengths ($b_{i}^{(1)}>0$, $b_{i}^{(2)}>0$, $b_{i}^{(3)}>0$) are 
distributed according a truncated Multivariate Normal (MVN)
with mean vector, $\mathbf{\mu}$, of length 3 and a $3 \times 3$ covariance matrix, $\Sigma$, conditional on the hidden state $H_{a}^{i}$:
\begin{equation}
\label{eq:bond_density}
(b_{i}^{(1)}, b_{i}^{(2)}, b_{i}^{(3)})  \sim \text{MVN}\big(\mu(H_a^{i}),\,\Sigma(H_a^{i})\big)
\end{equation}
Note that the values of the three bond angles and three bond lengths are all largely invariant, implying that the values can be reasonably fixed. However, we opted to treat them as random variables so that the model gives a complete probabilistic description of a protein backbone structure.

\paragraph{Structural likelihood}
The likelihood of a structural observation, $p\big(X_{a}^{i}\,|\,H_{a}^{i},S_{a}^{i},\theta\big)$, at site $i$ conditional on the hidden state, $H^{i}_a$, and amino acid, $S^{i}_a$, is given by a product of the densities in Equations~\ref{eq:dihedral_density}, \ref{eq:angle_density}, and \ref{eq:bond_density}.

\paragraph{Hidden Markov model}
The joint likelihood, $\Pi(a)$, of the structural observations, $X_a$, and hidden states, $H_a$  conditioned on the amino acid observations, $S_a$, is given by a hidden Markov model as follows:
\begin{align}
\label{eq:proteinlikelihood}
\Pi(a) = p(H_{a},X_{a}\,|\,S_{a},\theta) = \notag \\
 p\big(X_{a}^{1}\,|\,H_{a}^{1},S_{a}^{1},\theta\big)p(H_{a}^{1},\theta)\notag \\
\times \prod_{i=2}^n p\big(X_{a}^{i}\,|\,H_{a}^{i},S_{a}^{i},\theta\big)p(H_{a}^{i}\,|\,H_{a}^{i-1},\theta) 
\end{align}
where $p(H_{a}^{i}\,|\,H_{a}^{i-1},\theta)$ is a transition probability matrix between hidden states intended to capture the amino acids (along with their associated structural values) that tend to occur next to one another in the peptide backbone.

\begin{comment}
The probability, $\Pi(a)$, of a set structural observations, $X_a$ conditioned on the corresponding hiddem state sequence, $H_a$,  the amino acid sequence, $S_a$, and model parameters, $\theta$, is given by:
\begin{align}
\Pi(a) = p(X_{a}\,|\,H_{a},S_{a},\theta) = \notag \\
\frac{1}{Z} p\big(X_{a}^{i}\,|\,H_{a}^{1},S_{a}^{1},\theta\big)p(H_{a}^{1},\theta)\notag \\
\times \prod_{i=2}^n p\big(X_{a}^{i}\,|\,H_{a}^{i},S_{a}^{i},\theta\big)p(H_{a}^{i}\,|\,H_{a}^{i-1},\theta) 
\end{align}

where the normalising constant:
\begin{align}
Z = \,\sum_{H} \big[ p\big(X_{a}^{i}\,|\,H_{a}^{1},S_{a}^{1},\theta\big)p(H_{a}^{1},\theta)\notag \\
\times \prod_{i=2}^n p\big(X_{a}^{i}\,|\,H_{a}^{i},S_{a}^{i},\theta\big)p(H_{a}^{i}\,|\,H_{a}^{i-1},\theta) \big]
\end{align}
where the sum runs over all possible hidden state sequences and can be calculated in $\mathcal{O}(h^2|P_{a}|)$ computational steps using the HMM forward algorithm.
\end{comment}

\subsubsection{Evolutionary model}
Thus far we have only considered a single protein. In this section we outline how multiple phylogenetically related proteins are modelled evolutionarily. Following \citet{choi2008basing} we construct a rate matrix that represents changes between two sequences, $a$ and $b$, instead of character states, such as amino acids, as is typical of substitution models. Furthermore, each sequence position combines a hidden state ($h_i$), an amino acid ($aa_i$), and a set of structural observations ($x_i$), into a joint character state ($h^{s}_i,aa^{s}_i,x^{s}_i$), where $s$ refers to a particular sequence. The rate matrix is given as follows:
\begin{equation}
\label{eq:ratematrix}
M_{ab}=\begin{cases}

\sqrt{\frac{\Pi(b)}{\Pi(a)}}S_{a_{i}b_{i}}\pi^{h^{b}_{i}}_{aa^{b}_{i}} & \text{Single amino acid}\\
& \text{difference at site $i$.}\\ 

\sqrt{\frac{\Pi(b)}{\Pi(a)}}R_{a_{i}b_{i}}\pi^{h^{b}_{i}}_{aa^{b}_{i}} & \text{Single hidden state}\\
& \text{difference at site $i$.}\\ 

0 & \text{Both hidden state and}\\
& \text{amino acid differences}\\
& \text{at site $i$.}\\ 

0 & \text{Differences at two}\\
& \text{or more sites.}\\

-\underset{c\neq a}{\sum}M_{ac} & a=b
\end{cases}
\end{equation}

where $i$ is the position that differs between sequence $a$ and $b$, $S$ is a symmetric $20\times{20}$ amino acid exchangeability matrix, and $R$ is a symmetric $h\times{h}$ hidden state exchangeability matrix. The term, $\frac{\Pi(b)}{\Pi(a)}$, weights each hidden state or amino acid change, such that sequences are visited with probability given by it's probability under the hidden Markov model multiplied by the amino acid sequence probability. This gives an evolutionary model on amino acid sequences and protein structures that accounts for neighbouring dependencies between adjacent sites and introduces temporal evolutionary dependencies between proteins.

The evolutionary dependencies between structures is introduced via the hidden states, thus avoiding having to directly implement an evolutionary process on structure, which is cumbersome given the continuous nature of the structural observations. We have previously developed a continuous diffusion process on angles for modelling protein dihedral angles, \citep{Garcia-Portugues:ads, Golden2017}, however,  protein backbone angles and bond lengths do not evolve in a continuous fashion, rather they are expected to 'jump' when changes occur. The hidden states capture this jump behaviour.
 
Note that proteins $P_a$ and $P_b$ referred to in the ratio $\frac{\Pi(b)}{\Pi(a)}$ in Equation~\ref{eq:ratematrix} always differ at exactly one site, implying that at most three terms in Equation~\ref{eq:proteinlikelihood} need to be considered when computing the ratio. 

Additionally note, although the summation in Equation~\ref{eq:ratematrix} appears to involve an exponential number of terms, most terms are equal to zero, except those that differ from $P_a$ at one position. Furthermore, an amino acid transition and a hidden state transition are not permitted to occur simulatenously, further reducing the number of terms that are summed ($19n$ amino acid terms plus $(h-1)n$ hidden state terms).


\paragraph{Stationary probability of proteins}
Following \citet{choi2008basing}, and by construction, the stationary probability of a protein $a$ is given by:
\begin{equation}
\label{eq:stationary_dist}
p(P_a|\theta) = \frac{ \Pi(a) \underset{i}{\prod}\pi^{h^{a}_{i}}_{aa^{a}_{i}} }{ \sum_{k} \Pi(k) \underset{i}{\prod}\pi^{h^{k}_{i}}_{aa^{k}_{i}} } 
\end{equation}


\paragraph{Time-reversibility}
Since $S$ and $R$ are symmetric matrices, i.e. $S_{a_{i}b_{i}}=S_{b_{i}a_{i}}$ and $R_{a_{i}b_{i}}=R_{b_{i}a_{i}}$, it can easily be shown that time reversibility holds, in other words:
\begin{equation}
p(P_a|\theta)M_{ab} = p(P_b|\theta)M_{ba}
\end{equation}


\paragraph{Dataset likelihood}
\begin{align}
p(\mathcal{D}_d|\mathcal{T}_d,\theta)=\notag\\
\prod_{b \in \mathcal{T}_d} p\big(X_{s}^{i}(t_{end})\,|\,H_{s}^{i}(t_{end}),S_{s}^{i}(t_{end}),\theta\big)\notag\\ 
\times e^{M_{b_{n}}(t_{end}-t_{n})} \big[\prod_{k=1}^{n}  e^{-M_{b_{k}b_{k}}(t_{k}-t_{k-1})} M_{b_{k-1}b_{k}} \big]
\end{align}

\subsection{Model training}

\subsubsection{Datasets}

\subsubsection{Parameter estimation}
Stochastic EM (StEM, \citet{gilks1995markov}) was used to train the model. StEM  is a stochastic version of the well known Expectation-Maximization algorithm \citep{gilks1995markov}. Its distinguishing feature is that the E-step consists of filling in the values of the latent variables using sampling. Only a single value is sampled. StEM is attractive due to its computational efficiency and its tendency to avoid getting stuck in local minima \citep{gilks1995markov}.

Sampling was used in the E-step to sample branch paths and times. In other words, at iteration $k$ for each dataset, $d$, consisting of an aligned set of proteins, $\mathcal{D}_d$,  and a corresponding phylogenetic tree, $\mathcal{T}_d$, we draw samples, from the following joint-distribution:
\begin{align*}
Z_{d}^{(k)}\sim p(\mathcal{T}_d|\mathcal{D}_d,\Psi^{(k)}).
\end{align*}

In the M-step the samples from the previous E-step, were used to update the hidden node parameters ($\hat{\Psi}$) using efficient sufficient statistics (ESSs).
\subsection{Inference}



\subsubsection{Gibbs sampling}


\begin{table*}
	\captionsetup{justification=centering}
	\caption{\label{tab:sequencereconstruction} Ancestral sequence reconstruction benchmarks}	
	\begin{tabularx}{1.0\linewidth}{ccccc}
	\toprule
	Dataset & Our model & LG2008 & BEAST & ASR\\
	\midrule
	\rowcolor{black!20} Influenza & 0.896 & 0.876 & 0.88 & 0.87\tabularnewline
	HIV & 0.896 & 0.876 & 0.88 & 0.87\tabularnewline
	\bottomrule
	\end{tabularx}
\end{table*}


%%% Results and discussion %%%
\section{Results and Discussion}

\subsection{Benchmarks of ancestral sequence reconstruction}
	
%%% Conclusions %%%
\section{Conclusions}

%%% Software availability %%%
\section{Software availability}
Julia code (compatible with Windows and Linux) is available at: \href{https://github.com/michaelgoldendev/MESSI}{https://github.com/michaelgoldendev/MESSI}

%%% Acknowledgements %%%
\section{Acknowledgements}
MG is supported by the ERC under the European Union’s Seventh Framework Programme (FP7/2007-2013)/ERC grant agreement no. 614725-PATHPHYLODYN. 

\ifmbeformat
\section{Supplementary material}
Supplementary material is available  at Molecular Biology and Evolution
online: \url{http://www.mbe.oxfordjournals.org/}
\fi

\bibliographystyle{natbib}%%%%natbib.sty
\bibliography{refs}%%%refs.bib


\end{document}